llvm/lib/Transforms/Utils/LoopPeel.cpp

```diff
@@ -636,9 +636,13 @@ static void updateBranchWeights(Instruction *Term, WeightInfo &Info) {
                     MDB.createBranchWeights(Info.Weights));
   for (auto [Idx, SubWeight] : enumerate(Info.SubWeights))
     if (SubWeight != 0)
-      Info.Weights[Idx] = Info.Weights[Idx] > SubWeight
-                              ? Info.Weights[Idx] - SubWeight
-                              : 1;
+      // Don't set the probability of taking the edge from latch to loop header
+      // to less than 1, as this could significantly reduce the loop's hotness,
+      // which would be incorrect in the case of underestimating the trip count.
+      Info.Weights[Idx] =
+          Info.Weights[Idx] > SubWeight
+              ? std::max(Info.Weights[Idx] - SubWeight, SubWeight)
+              : SubWeight;
```

Comment:
```suggestion
      Info.Weights[Idx] =
          Info.Weights[Idx] > 2 * SubWeight
              ? Info.Weights[Idx] - SubWeight
              : SubWeight;
```

Reply:
Actually this suggestion may be susceptible to overflow... So please keep your original `std::max` variant of the code then and ignore this!